debate_app.py

```diff
@@ -1,128 +1,314 @@
-#!/usr/bin/env python3
-
 import time
+import os
+from typing import Dict, List, Tuple, Optional
 from rich.console import Console
 from rich.panel import Panel
 from rich.text import Text
 from rich.prompt import Prompt
 from rich.progress import Progress, SpinnerColumn, TextColumn
-from personality import PersonalityConfig, create_personality
+from rich.columns import Columns
+from personality import PersonalityConfig, create_personality, LLMPersonality
 
 console = Console()
 
-
 class DebateApp:
+    """Main application class for running AI-powered debates.
+    
+    This class manages the debate process, including selecting AI providers,
+    managing personalities, and orchestrating the debate rounds.
+    """
+    
     def __init__(self):
-        self.personalities = self._create_personalities()
-        self.judge = self._create_judge()
 
-    def _create_personalities(self):
+        """Initialize the debate application and setup required components."""
+        self.selected_providers: List[str] = self._select_providers()
+        while True:
+            try:
+                self._check_and_request_api_keys()
+                self.personalities: Dict[str, LLMPersonality] = self._create_personalities()
+                break
+            except ValueError as e:
+                console.print(f"\n[bold red]Error:[/bold red] {str(e)}")
+                retry = Prompt.ask(
+                    "\n[yellow]Would you like to try entering the API keys again?[/yellow]",
+                    choices=["y", "n"]
+                )
+                if retry.lower() != "y":
+                    raise SystemExit("Exiting due to API key validation failure")
+        self.judge: Optional[LLMPersonality] = None
+        
+    def _check_and_request_api_keys(self) -> None:
```

Comment:
Can this be a separate file? Maybe name it `api_keys_utils.py` as the app code is getting complex